# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
# import numpy and pandas
import pandas as pd
import numpy as np
from scipy.stats import trim_mean, mode, skew, gaussian_kde, pearsonr, spearmanr, beta
from statsmodels.stats.weightstats import ztest as ztest

from scipy.stats import ttest_ind, norm, t
from scipy.stats import f_oneway
from scipy.stats import sem

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [2]:
# Run this code:
salaries = pd.read_csv('../data/Current_Employee_Names__Salaries__and_Position_Titles.csv')

Examine the `salaries` dataset using the `head` function below.

In [3]:
salaries.head()

,Name,Job Titles,Department,Full or Part-Time,Salary or Hourly,Typical Hours,Annual Salary,Hourly Rate
0,"AARON, JEFFERY M",SERGEANT,POLICE,F,Salary,NaN,101442.0,NaN
1,"AARON, KARINA",POLICE OFFICER (ASSIGNED AS DETECTIVE),POLICE,F,Salary,NaN,94122.0,NaN
2,"AARON, KIMBERLEI R",CHIEF CONTRACT EXPEDITER,GENERAL SERVICES,F,Salary,NaN,101592.0,NaN
3,"ABAD JR, VICENTE M",CIVIL ENGINEER IV,WATER MGMNT,F,Salary,NaN,110064.0,NaN
4,"ABASCAL, REECE E",TRAFFIC CONTROL AIDE-HOURLY,OEMC,P,Hourly,20.0,NaN,19.86


In [7]:
salaries['Hourly Rate'].value_counts()

Hourly Rate
35.60    1415
40.20     781
36.21     626
19.86     418
46.10     306
         ... 
42.25       1
32.59       1
49.52       1
28.00       1
35.43       1
Name: count, Length: 167, dtype: int64

In [10]:
salaries['Hourly Rate'].isna.sum()

AttributeError: 'function' object has no attribute 'sum'

In [8]:
len(salaries)

33183

In [9]:
salaries['Hourly Rate'].describe()

count    8022.000000
mean       32.788558
std        12.112573
min         2.650000
25%        21.200000
50%        35.600000
75%        40.200000
max       109.000000
Name: Hourly Rate, dtype: float64

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [13]:
from scipy.stats import ttest_1samp

In [23]:
wage_sample = salaries[salaries['Salary or Hourly']=='Hourly']['Hourly Rate'].dropna()

t_stat, p_val = ttest_1samp(wage_sample, 30)

print(f"T-Statistic: {t_stat}")
print(f"P-Value: {p_val}")

T-Statistic: 20.6198057854942
P-Value: 4.323024048623005e-92


In [ ]:
#T-Distribution table

# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [16]:
from scipy.stats import sem

In [19]:
n = len(wage_sample)
mean_hr = wage_sample.mean()
std_err = sem(wage_sample)

#95% Confidence Interval
ci = t.interval(0.95, df=n-1, loc=mean_hr, scale=std_err)

print(f"The 95% confidence Interval: {ci}")

The 95% confidence Interval: (np.float64(32.52345834488425), np.float64(33.05365708767623))


# Challenge 4 - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [20]:
from statsmodels.stats.proportion import proportions_ztest

In [22]:
count_hourly = len(salaries[salaries['Salary or Hourly'] == 'Hourly'])
nobs = len(salaries)

z_stat, p_val_prop = proportions_ztest(count_hourly, nobs, value=.25)

print(f"Sample Proportion: {count_hourly/nobs:.4f}")
print(f"Z-statistic: {z_stat}")
print(f"P-Value: {p_val_prop}")

Sample Proportion: 0.2418
Z-statistic: -3.5099964213703005
P-Value: 0.0004481127249057967


"The number of hourly workers in the City of Chicago is 24.18%, which is statistically significantly different from the hypothesized 25%. This is confirmed by a p-value of 0.000448, which is well below the significance level of 0.05. Therefore, we reject the null hypothesis."